### PR TITLE
Add MiMa checks and fix binary incompatible changes

### LINKED ISSUE #1702: Setup MiMa to prevent unintended breaking changes
Now that Caliban has the versioning scheme defined as `pvp`, it would be good to setup mima checks during CI so that we can catch any unintended breaking changes.

### PR DESCRIPTION
**Pull request by kyri-petrou**
This PR adds MiMa checks during CI. See comments below for some more info. Just some points for discussion:

1. Should we be strict when the code is source compatible but not binary compatible? (e.g., the `makeScalar` method wasn't bincompat)
2. What's the best way to ensure that we clear the MiMa filters after each release?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -80,7 +80,7 @@ jobs:
       - checkout
       - restore_cache:
           key: sbtcache
-      - run: sbt ++3.2.2 core/test catsInterop/compile benchmarks/compile monixInterop/compile clientJVM/test clientJS/compile zioHttp/test tapirInterop/test http4s/test federation/test reporting/test tracing/test
+      - run: sbt ++3.2.2 core/test catsInterop/compile benchmarks/compile monixInterop/compile clientJVM/test clientJS/compile zioHttp/test tapirInterop/test http4s/test federation/test tools/test reporting/test tracing/test
```
Comment (kyri-petrou) [inline .circleci/config.yml:86]: We were only testing `tools` against Scala 2.12. Was that intentional?

File: core/src/main/scala/caliban/schema/Types.scala [hunk 2]
```diff
@@ -7,14 +7,19 @@ import scala.annotation.tailrec
...
-    directives: Option[List[Directive]] = None
+    specifiedBy: Option[String] = None
+  ): __Type =
+    makeScalar(name, description, specifiedBy, None)
+
+  def makeScalar(
+    name: String,
+    description: Option[String],
+    specifiedBy: Option[String],
+    directives: Option[List[Directive]]
```
Comment (kyri-petrou) [inline core/src/main/scala/caliban/schema/Types.scala]: So this change was source compatible but not binary compatible. I'm 50-50 whether I should revert this change and add this method as a filter instead

File: build.sbt [hunk 3]
```diff
@@ -579,6 +595,22 @@ lazy val commonSettings = Def.settings(
...
 )
 
+lazy val enableMimaSettingsJVM =
+  Def.settings(
+    mimaPreviousArtifacts := previousStableVersion.value.map(organization.value %% moduleName.value % _).toSet,
+    mimaBinaryIssueFilters ++= Seq( // Filters should be cleared after a new release
+      ProblemFilters.exclude[MissingTypesProblem]("caliban.schema.SchemaDerivation$Auto"),
+      ProblemFilters.exclude[ReversedMissingMethodProblem](
+        "caliban.schema.SchemaDerivation#Auto.caliban$schema$SchemaDerivation$Auto$$$outer"
+      )
```
Comment (kyri-petrou) [inline build.sbt]: The changes to `Schema.Auto` are not bincompat but that's okay since it wasn't working as intended previously

File: core/src/main/scala/caliban/schema/Types.scala (hunk 2 above)
Comment (ghostdogpr) [reply core/src/main/scala/caliban/schema/Types.scala]: Considering the changes I'm working on with the adapters, I think we may go 2.2 directly so maybe it's fine to keep those breaking changes.

File: .circleci/config.yml (hunk 1 above)
Comment (ghostdogpr) [reply .circleci/config.yml:86]: Probably an oversight 👍

**Comment by ghostdogpr** [comment]
> 1. Should we be strict when the code is source compatible but not binary compatible? (e.g., the `makeScalar` method wasn't bincompat)

I think we should be strict for patch versions.

> 2. What's the best way to ensure that we clear the MiMa filters after each release?

I have a step where I update the version in docs, I can do it at the same time (it's manual though).

How do we make Mima not fail when we know the next release will be a minor version release (in which case, we allow breaking changes)?

File: core/src/main/scala/caliban/schema/Types.scala (hunk 2 above)
Comment (kyri-petrou) [reply core/src/main/scala/caliban/schema/Types.scala]: > How do we make Mima not fail when we know the next release will be a minor version release (in which case, we allow breaking changes)?

@ghostdogpr I  added a flag in `build.sbt` to disable CI breaking on binary incompatible changes. The good thing about this flag is that the plugin will still report the bincompat issues in the console, so we can always look at it when we're close to releasing to know what bin / source incompatible changes we introduced

Also, Since the next release will likely be 2.2.0, I reverted the changes to fix the bincompat issue in `makeScalar`.

**Comment by kyri-petrou** [comment]
> How do we make Mima not fail when we know the next release will be a minor version release (in which case, we allow breaking changes)?

@ghostdogpr I  added a flag in `build.sbt` to disable CI breaking on binary incompatible changes. The good thing about this flag is that the plugin will still report the bincompat issues in the console, so we can always look at it when we're close to releasing to know what bin / source incompatible changes we introduced

Also, Since the next release will likely be 2.2.0, I reverted the changes to fix the bincompat issue in `makeScalar`.

File: core/src/main/scala/caliban/introspection/adt/__Directive.scala [hunk 4]
```diff
@@ -17,4 +17,7 @@ case class __Directive(
       isRepeatable,
       locations.map(_.toDirectiveLocation)
     )
+
+  @deprecated("Use isRepeatable")
+  def repeatable: Boolean = isRepeatable
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/introspection/adt/__Directive.scala]: Can we delete these then?

File: core/src/main/scala/caliban/schema/Schema.scala [hunk 5]
```diff
@@ -147,6 +147,15 @@ trait GenericSchema[R] extends SchemaDerivation[R] with TemporalSchema {
       override def resolve(value: A): Step[Any]                              = PureStep(makeResponse(value))
     }
 
+  @deprecated("use the overloaded method with directives instead")
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/schema/Schema.scala]: Delete?

File: core/src/main/scala/caliban/schema/Types.scala [hunk 6]
```diff
@@ -7,9 +7,6 @@ import scala.annotation.tailrec
 
 object Types {
 
-  /**
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/schema/Types.scala]: Any reason to remove this?

File: core/src/main/scala/caliban/introspection/adt/__Directive.scala (hunk 4 above)
Comment (kyri-petrou) [reply core/src/main/scala/caliban/introspection/adt/__Directive.scala]: Done 👍 My bad, had to head out in a hurry earlier and didn't cleanup properly

File: core/src/main/scala/caliban/schema/Types.scala (hunk 6 above)
Comment (kyri-petrou) [reply core/src/main/scala/caliban/schema/Types.scala]: I deleted the extra method but didn't realise I also deleted the scaladoc 😅

**Review by ghostdogpr — APPROVED** [review]
